pkg/cmd/render/render.go

```diff
@@ -80,28 +136,98 @@ func (r *renderOpts) Complete() error {
 type TemplateData struct {
 	genericrenderoptions.ManifestConfig
 	genericrenderoptions.FileConfig
+
+	clusterInfrastructure  configv1.Infrastructure
+	cloudProviderConfigMap corev1.ConfigMap
 }
 
 // Run contains the logic of the render command.
 func (r *renderOpts) Run() error {
 	renderConfig := TemplateData{}
+
+	// XXX: What is this file used for?
 	if len(r.clusterConfigFile) > 0 {
 		_, err := ioutil.ReadFile(r.clusterConfigFile)
 		if err != nil {
 			return err
 		}
 		// TODO I'm thinking we parse this into a map and reference it that way
 	}
+
+	if len(r.clusterInfrastructureFile) > 0 && len(r.cloudProviderConfigInputFile) > 0 && len(r.cloudProviderConfigOutputFile) > 0 {
+
+		// Read, parse, and save the infrastructure object
+		var clusterInfrastructure configv1.Infrastructure
+		clusterInfrastructureFileData, err := ioutil.ReadFile(r.clusterInfrastructureFile)
+		if err != nil {
+			return err
+		}
+		if len(clusterInfrastructureFileData) > 0 {
+			err := yaml.Unmarshal(clusterInfrastructureFileData, &clusterInfrastructure)
+			if err != nil {
+				return err
+			}
+		}
+		renderConfig.clusterInfrastructure = clusterInfrastructure
+
+		// Read, parse, and save the user provided cloud configmap
+		var cloudProviderConfigInput corev1.ConfigMap
+		cloudProviderConfigInputFileData, err := ioutil.ReadFile(r.cloudProviderConfigInputFile)
+		if err != nil {
+			return err
+		}
+		if len(cloudProviderConfigInputFileData) > 0 {
+			err := yaml.Unmarshal(cloudProviderConfigInputFileData, &cloudProviderConfigInput)
+			if err != nil {
+				return err
+			}
+		}
+		renderConfig.cloudProviderConfigMap = cloudProviderConfigInput
+
+		// Determine the platform type
+		var platformName configv1.PlatformType
+		if pstatus := clusterInfrastructure.Status.PlatformStatus; pstatus != nil {
+			platformName = pstatus.Type
+		}
+		if len(platformName) == 0 {
+			platformName = clusterInfrastructure.Status.Platform
+		}
+
+		// Determine the platform specific transformer method to use
+		cloudConfigTransformers := kubecloudconfig.CloudConfigTransformers()
+		cloudConfigTransformerFn, ok := cloudConfigTransformers[platformName]
+		if !ok {
+			cloudConfigTransformerFn = kubecloudconfig.DefaultCloudConfigTransformer()
+		}
+		target, err := cloudConfigTransformerFn(&cloudProviderConfigInput, clusterInfrastructure.Spec.CloudConfig.Key, &clusterInfrastructure)
+		if err != nil {
+			return err
+		}
+
+		targetCloudConfigMap := kubecloudconfig.TargetConfigName()
+
+		target.Name = kubecloudconfig.TargetConfigName()
+		target.Namespace = operatorclient.GlobalMachineSpecifiedConfigNamespace
+		/* ApplyConfigMap() */
+
+		targetCloudConfigMapData, err := yaml.Marshal(target)
+		if err != nil {
+			return err
+		}
```

Comment:
all this should be part of the kube_cloud_config package.

we can keep the load inputs in here, but how/when to generate the new cloud config file should be controlled by kube_cloud_config package. the sync also needs something like that so maybe we can re-use one implementation...

Reply:
I've moved the code to kube_cloud_config. I still need to look at sync() and perhaps generalize the transformer functions also.